### ISSUE TITLE
Flexible error response control

### ISSUE THREAD

**Issue by KUTlime** [FastEndpoints/FastEndpoints#771]
I'm relatively new to FastEndpoints, but I consider it a very mature library. Currently, I'm struggling with an elegant way how to generate error responses for various types of situations that can occur.

I have a mediator command that returns a discriminated union (_a coproduct_) generated using `OneOf` library. This keeps my API features flexible.

![image](https://github.com/user-attachments/assets/e0c3ae96-b73d-4cce-aaa0-999edd9b714b)

I'm looking for a way how to simplify this code:

```csharp
public override async Task HandleAsync(UpdateConfigurationRecordRequest request, CancellationToken ct)
{
    var result = await mediator.Send(request.ToCommand(), ct).ConfigureAwait(continueOnCapturedContext: false);
    await result.Match(
        cr => SendOkAsync(cr.ToDto(), cancellation: ct).ConfigureAwait(ConfigureAwaitOptions.None),
        notFound => SendNotFoundAsync(ct).ConfigureAwait(ConfigureAwaitOptions.None),
        deviceNameInConflictError => HttpContext.Response.SendAsync($"The incoming Device name '{deviceNameInConflictError.UpdateDeviceName}' doesn't correspond to the record Device name '{deviceNameInConflictError.RecordDeviceName}'", statusCode: 409, cancellation: ct).ConfigureAwait(ConfigureAwaitOptions.None),
        configurationLabelNotFoundError => HttpContext.Response.SendAsync($"The configuration label with ID: '{configurationLabelNotFoundError.Id}' not found.", statusCode: 404, cancellation: ct).ConfigureAwait(ConfigureAwaitOptions.None),
        configurationVersionNotFoundError => HttpContext.Response.SendAsync($"The configuration version with ID: '{configurationVersionNotFoundError.Id}' not found.", statusCode: 404, cancellation: ct).ConfigureAwait(ConfigureAwaitOptions.None),
        e => SendErrorsAsync(statusCode: 500, ct).ConfigureAwait(ConfigureAwaitOptions.None));
}
```

I can't use `SendAsync()` directly, and I can't use `SendNotFoundAsync(...)`, because it doesn't provide an overload with error details fast through, the same goes for `SendErrorsAsync(...)`.

Is there any way how to do it elegantly, or bring some support for typed unions pattern support?

**Comment by dj-nitehawk** [comment]
have you had a look at the "results pattern" section of the [cookbook](https://fast-endpoints.com/docs/the-cookbook#results-pattern)?

there's no example there for the `OneOf` package but i think you can cook something up by looking at those examples.

let me know if that doesn't help and i'll have a deeper look.

**Comment by KUTlime** [comment]
Hi @dj-nitehawk, once again, thank you for this library and the excellent support that you giving to us. 🫱🏿‍🫲🏿

## Short version

Some new Send methods would be nice, like SendNotFound with in-value overload, SendConflict(...), etc. to support discriminated-union-based output from a service, or a mediator.

## TL;DR

I did. I "cooked" my code based on the samples you mentioned.

The [Ardalis.Result](https://gist.github.com/dj-nitehawk/c209274f6bfcc7436d8f463f8ace554a) gist isn't a good way to do it.

The discriminated unions could be different between endpoints, so the extension method won't be very useful, or it would lead to violations of DRY, but OK, with some clear use of generic, this could be bypassed (_IMHO: FE could be much helpful_).

The gist based on [ErrorOr](https://gist.github.com/dj-nitehawk/66cba78a1a3a1e0799d87d67d8aa14bd) is just overengineering to me that could be avoided by extending the send methods.

It is very nice to have granular control over configuration, auto validation, etc., but if it isn't supported with corresponding granular yet simple control over return values driven by some simple logic in the endpoint, ehm... 🤷🏿‍♀️🥹

Yes, there is the [Union-Type returning handler](https://fast-endpoints.com/docs/get-started#union-type-returning-handler) which is most likely the right way to go (_at least in my case_), but is it following RFC 7807? Does the code `FastEndpoints.ProblemDetails(ValidationFailures);` follow RFC 7807? Will it be consistent for the client with the lines like `return TypedResults.NotFound();` which returns, if I'm not mistaken, the [`ProblemDetails`](https://learn.microsoft.com/en-us/dotnet/api/microsoft.aspnetcore.mvc.problemdetails?view=aspnetcore-8.0) instance to a client?

**Comment by dj-nitehawk** [comment]
![image](https://github.com/user-attachments/assets/5be1ad30-99cb-4f47-b298-7e0c129027e2)

custom extension methods like above is the recommended approach for creating custom Send methods.
if having `this.` in your code is undesirable, the next approach is to inherit the `Endpoint<TRequest,TResponse>` class and add your own Send methods and whatever else you need.

 ---
 
 `FastEndpoints.ProblemDetails` class is `RFC7807` [compatible](/docs/configuration-settings#rfc7807-rfc9457-compatible-problem-details).
 
 and you can do `TypedResults.NotFound(new FastEndpoints.ProblemDetails());` because `FastEndpoints.ProblemDetails` is an `IResult` type.
 
 you can also do the following:
 
 ```cs
 SendResultAsync(new FastEndpoints.ProblemDetails());
 ```
 
 ---
 
 i'll investigate what we can do to improve union type support but it's going to be difficult to have a common approach for all union type libraries out there. most likely the situation would improve when microsoft releases their discriminated union c# support.

**Comment by nhwilly** [comment]
Here's my experience.

I use `Ardalis.Result` extensively.  But there will *never* be a one-to-one match for `HttpStatusCode` or FE extension methods.  After looking at the suggestions, I ended up doing global response sender from in the examples.  Really great addition, FE folks.

I take the `Result<TDto>`, catch it in the Global sender and ensure that validation errors (from Result) and validation failures (from FluentValidation) are handled and I map the lot of it to a ProblemDetails.  Because I'm using Blazor WASM and because I didn't want to include the *entire* library as a dependency (to be downloaded a zillion times) I just copied the `ProblemDetails` from MS as an abstract class and then derived my own from it.  Of course I added `Error[]` and `ValidationError[]` properties.

Full disclosure, the Microsoft ProblemDetails can't be deserialized without a converter (at least as I remember), so I added one of those, too.  My derived class came later, so I'm not sure I need the converter anymore...

In the end, this means that all my endpoints issue `Result<TDto>` values and I map them to the appropriate `send` command.

And on the client side, I can use the same custom problem details to handle *anything* other than the successful response by passing the `Result` up the stack.  Works for me.  I'm sure others will need something different.

HTH-YMMV
